### ISSUE TITLE
[BUG] Photon Conservation not implemented for lower-level functions

### ISSUE THREAD

**Issue by steven-murray** [21cmfast/21cmFAST#118]
**Describe the bug:**

Currently the PHOTON_CONS flag is only useable for `run_coeval` and `run_lightcone`. If the user attempted to call `ionized_box` it'll fail. Perhaps `ionized_box` should also have the ability to internally setup the photon non-conservation correction? In doing so, I suspect that might be a little more complicated.

**Expected behavior:**

One should be able to call `ionized_box` with `PHOTON_CONS` and it should all just work.

**Comment by BradGreig** [comment]
In terms of getting it to work directly from `ionized_box` I don't think its too much work. The worry of course is the overlap between that and calling `ionized_box` from within `run_coeval` and `run_lightcone` (where `ionized_box` is called multiple times).

I guess a similar solution to #114 could work. Where a check is performed to see if the initialisation of the photon non-conservation has been performed already. This ensures that if the user is performing `ionized_box` multiple times, or if `run_coeval` and `run_lightcone` are used that the initialisation and setup is only performed once.

**Comment by steven-murray** [comment]
Yeah, I think if we implement the suggestion from #114, then this shouldn't be too hard (especially if that ps struct is made available to python to check).

**Comment by steven-murray** [comment]
To bump this in importance -- I just discovered a bug in which there was double-freeing of the photon cons structures when computing a coeval then a lightcone (though it would been the same for any pair of high-level objects), with the same parameters. This is because the `IonisationBox` function is not run for the second object at all -- all the values are merely cached. The actual allocation of the photon cons is happening *inside* the IonisationBox routine. Thus they are being free'd at the end without being allocated. 

I've fixed it by ensuring that we don't try to free if a global `photon_cons_allocated` variable is not True. However, we can't easily go back and get that photon cons data, even though it's technically been used in the lightcone. This speaks to the need (in my opinion) for the photon cons routines to return structs that can be initialized via python or C, and are _not_ global, but are passed around like anything else.

**Comment by BradGreig** [comment]
Hey @steven-murray, yeah, I agree that it makes more sense to pass around the photon cons data through structs. I think the best solution is to have the photon cons data writing to its own unique file (with hash, like initial conditions etc.). The reason for this is that it cannot be attached to any other file as it'll just be duplicated effort.

This photon cons struct would then _only_ need to be passed to the ionised box function, which should then have a check if file exists or create a new one (allowing photon cons to be easily added to the lower level function).

I can create a branch and put up a draft PR for you to track the implementation to ensure this is performed properly. Just give me the go ahead and I'll get to work on it.

**Comment by steven-murray** [comment]
Hey Brad, yup, that sounds exactly like the right idea to me. I would be very happy for you start a PR on it. I reckon we leave it to 3.1 to merge in though (I wanna get this paper out!)
